lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+		return nil
+	}
+
+	if unflushedBytes.StateChangeBytes == nil || unflushedBytes.StateChangeOperationIndexes == nil || len(unflushedBytes.StateChangeOperationIndexes) == 0 {
+		return fmt.Errorf("Error flushing state change file: FlushId %v has nil bytes\n", flushId)
+	}
+
+	// Write the encoded StateChangeEntry bytes to the state changer file.
+	_, err := flushFile.Write(unflushedBytes.StateChangeBytes)
+
+	if err != nil {
+		return fmt.Errorf("Error writing to state change file: %v", err)
+	}
+
+	// Buffer to hold bytes for index file
+	stateChangeIndexBuf := make([]byte, 0)
+
+	// Loop through the index bytes of the state change file and write them to the index file.
+	// The StateChangeOperationIndexes array contains the byte index of		 where each transaction occurs within the
+	// unflushed state change bytes (e.g. the first value in the slice will always be 0).
+	// We need to add the size of the state change file to each of these values to get the byte index
+	// in the state change file.
+	for _, indexBytes := range unflushedBytes.StateChangeOperationIndexes {
+		// Get the byte index of where this transaction occurs in the state change file.
+		dbOperationIndex := indexBytes + flushFileSize
+		// Convert the byte index value to a uint64 byte slice.
+		dbOperationIndexBytes := make([]byte, 8)
+		binary.LittleEndian.PutUint64(dbOperationIndexBytes, dbOperationIndex)
+
+		stateChangeIndexBuf = append(stateChangeIndexBuf, dbOperationIndexBytes...)
+	}
+
+	// Write the encoded uint32 indexes to the index file.
+	_, err = indexFile.Write(stateChangeIndexBuf)
+	if err != nil {
+		return fmt.Errorf("Error writing to state change index file: %v", err)
+	}
+	if event.IsMempoolFlush {
+		stateChangeSyncer.StateChangeMempoolFileSize += unflushedBytes.StateChangeBytesSize
+	} else {
+		stateChangeSyncer.StateChangeFileSize += unflushedBytes.StateChangeBytesSize
+	}
+
+	// Update unflushed bytes map to remove the flushed bytes.
+	delete(stateChangeSyncer.UnflushedBytes, flushId)
+	return nil
+}
+
+// SyncMempoolToStateSyncer flushes all mempool transactions to the db, capturing those state changes
+// in the mempool state change file. It also loops through all unconnected transactions and their associated
+// utxo ops and adds them to the mempool state change file.
+func (stateChangeSyncer *StateChangeSyncer) SyncMempoolToStateSyncer(server *Server) (bool, error) {
+	originalCommittedFlushId := stateChangeSyncer.BlockSyncFlushId
+
+	if server.mempool.stopped {
+		return true, nil
+	}
+
+	blockHeight := uint64(server.blockchain.bestChain[len(server.blockchain.bestChain)-1].Height)
+
+	stateChangeSyncer.BlockHeight = blockHeight
+
+	mempoolUtxoView, err := server.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		return false, errors.Wrapf(err, "StateChangeSyncer.SyncMempoolToStateSyncer: ")
+	}
+	// Kill the snapshot so that it doesn't affect the original snapshot.
+	mempoolUtxoView.Snapshot = nil
+	mempoolUtxoView.IsMempoolView = true
```

Comment:
OK I think this is going to be a super annoying comment, but let me run it by you.

I don't love this IsMempoolView field, and I was thinking about how we could remove it. The reason I don't like it is because it's kindof confusing on the UtxoView struct, and we have to thread it through *everywhere* even though it doesn't really change over the course of a flush.

So I had the following suggestion:
1. Move this field to the EventManager object, which we're already threading through everywhere.
2. Delete it from all of the arguments we're passing it to now. This is why this comment is annoying: It means you have to go to db_utils.go and change a lot of function signatures to remove the field...
3. To make SyncMempoolToStateSyncer work, create a new EventManager struct and call stateSyncerOperation stateSyncerFlush on that in this function rather than using the existing one.
4. Then I think you can remove the field from UtxoView and db_utils.go and wherever else.

I don't want it to be something that takes a whole day, though. Can you just look at how annoying it is and make a call? If it's more than an hour ish to do it then let's not worry about it for now.